lua_examples/pipeutils.lua

```diff
@@ -0,0 +1,54 @@
+-- A collection of pipe-based utility functions
+
+-- A convenience wrapper for chunking data arriving in bursts into more sizable
+-- blocks; `o` will be called once per chunk.  The `flush` method can be used to
+-- drain the internal buffer.  `flush` MUST be called at the end of the stream,
+-- **even if the stream is a multiple of the chunk size** due to internal
+-- buffering.  Flushing results in smaller chunk(s) being output, of course.
+local function chunker(o, csize, prio)
+  assert (type(o) == "function" and type(csize) == "number" and 1 <= csize)
+  local p = pipe.create(function(p)
+              -- wait until it looks very likely that read is going to succeed
+              -- and we won't have to unread.  This may hold slightly more than
+              -- a chunk in the underlying pipe object.
+              if 256 * (#p - 2) <= csize then return nil end
```

Comment:
What does #p return? That seems like an implementation dependent undocumented feature.
From the manual #userdata seems to return the size of the allocated block. But this usage implies a number of blocks.

Maybe pipe should get a dedicated size method.

Reply:
https://github.com/nodemcu/nodemcu-firmware/blob/4f6792773f93f36a7255cfd28dca7aa6c4aa9552/app/modules/pipe.c#L252 ; though I see it's not documented despite being implemented for Lua and not any internal reason (AFAICT).  I can add that.  Something like the following, maybe?

"The length (`__len`, `#`) of a `pipe` object is the number of records pending; this may be useful for estimating the total size of data in the `pipe`."

Reply:
Yes but what is a record in this case? Is it a memory chunk which contains 256 chars except for the last one? or is it something more complicated?. 
Maybe the __len implementation should be "fixed" to just return the remaining chars, not reveling the internal structure of the pipe.

Reply:
Reding the code it seems that the pipe object is a table which contains chunks of 256 bytes (LUAL_BUFFERSIZE), so #p would return the number of chunks in use +1 since the first index is used for the reader callback function.

The  `pipe__len` returns the used length of an individual chunk.

Maybe a __len method for the pipe itself would make sense, returning the total amount of buffered chars. 
As workaround, using the undocumented feature #p[n] (which returns the used length of chunk n) this code would return the exact amount of buffered chars. (haven't tested it, so it might not even compile)

```Lua
local len = 0
for i=2,#p,1 do
  len = len + #[i]
end
```

In general I would not want to use code which is not documented and relies on the implementation of something.

Reply:
There's no reason to compute an exact length here; it's just meant as a heuristic to guard the slower `:read()` below this line.  I'm just trying to avoid the case of, say, character-by-character `:write()`s causing quadratically many `:read(); :unread()` while building up the chunk. 

I think it makes sense to expose the "number of chunks", even if the chunks aren't full (which can happen after, say, `unread()`s).  The worst that happens is we end up causing the `pipe` object to consolidate several partial chunks into fuller chunks, which is probably a net benefit if it does happen, so long as it only happens occasionally.